### ISSUE TITLE
conch checks known_hosts for IP address but not hostname

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3494]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#3494|
|Type|enhancement|
|Created|2008-10-24 15:10:38Z|

I discovered this while working on [#10942](https://github.com/twisted/twisted/issues/10942), and likely it will be fixed by the same branch, but I want to track it as a separate ticket because it's actually a security issue.

When you do 'conch twistedmatrix.com', the client actually tries to verify the IP address, but _not_ the hostname, because `default.verifyHostKey` is passed the wrong thing.  This has been the case ever since [ticket:3059 the bug that caused `tcp.Client.getPeer` to return hostnames] was fixed; conch was depending on the fact that hostnames were returned.

There is actually an obscure attack against this: let's say you regularly ssh to two hosts, `foo.example.com` and `bar.example.com`, and an attacker knows this.  `bar.example.com` has serious security problems and you know this; `foo.example.com` does not.  You therefore routinely forward your SSH agent to `foo.example.com`.  If you use conch as your SSH client, then an attacker can exploit `bar.example.com`, poison your DNS cache so you think that `foo.example.com` has `bar.example.com`'s IP address, and get your agent forwarded to the host that is now under their control.  Thanks to the UI security issues with ssh agent forwarding, you probably won't notice when they start opening connections with your key.

<details><summary>Searchable metadata</summary>

```
trac-id__3494 3494
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1224861038000000 1224861038000000
changetime__1226804104000000 1226804104000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph set status to `closed`|
|-|-|

(In [25365]) Fixes for the conch client's known_hosts parsing and host key verification.

Fixes [#10942](https://github.com/twisted/twisted/issues/10942)
Fixes [#10890](https://github.com/twisted/twisted/issues/10890)
Fixes [#3494](https://github.com/twisted/twisted/issues/3494)
Fixes [#3496](https://github.com/twisted/twisted/issues/3496)
Fixes [#10884](https://github.com/twisted/twisted/issues/10884)
Fixes [#3499](https://github.com/twisted/twisted/issues/3499)

The main thrust of this change is that it adds a new module to conch, `twisted.conch.client.knownhosts`, which provides a structured representation of OpenSSH's `known_hosts` file.  This is a big step in the direction of modularizing conch's support for verifying host keys and storing the results of that verification, although the internal connection APIs are mostly unchanged at this point, they could now easily be adjusted to speak to an abstract interface and still manipulate the user's actual known_hosts entries rather than completely overriding them.

The individual bugs which this change fixes are too numerous to bother describing twice; have a look at the tickets themselves more information.  Suffice it to say that the conch client is now a lot more reliable (although it still has a long way to go).

In addition to making the conch client more stable, the new API allows for some programmatic manipulation of known_hosts files, which might be independently useful.

Additionally, although the review notes some ways in which the tests can still be improved, the tests for this module can be cited as a _much_ better example for conch contributors as to how unit tests within conch should look.  While these tests are covering almost exclusively new code, they are covering functionality which has existed (and been broken) for quite some time.

Author: glyph

Reviewers: exarkun, jml

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
